text/0000-new-test-waiters.md

```diff
@@ -0,0 +1,204 @@
+
+### `ITestWaiter`
+
+The `IWaiter` interface is built upon to create a more specific interface for a test waiter, `ITestWaiter`:
+
+```ts
+export interface ITestWaiter<T = Token> extends IWaiter {
+  beginAsync(token?: T, label?: string): T;
+  endAsync(token: T): void;
+  reset(): void;
+}
+```
+
+This interface is used for the two internal concrete types: `TestWaiter` and `NoopTestWaiter`. These types for the basis for the addon, and will likely satisfy the majority of use cases.
+
+The most common practice is to import and invoke the `buildWaiter` function to create a new test waiter. The recommendation is to do so at the module level, which allows a single waiter to be created per type. A single waiter is then usable across multiple instances.
+
+```ts
+function buildWaiter(name: string): ITestWaiter
+```
+
+In anything but a production build, this function will return a `TestWaiter` instance. When in production mode, a `NoopTestWaiter` will be returned, which as the name suggests is effectively a noop class. Since test waiters are intended to be called from application or addon code, but are only required to be _active_ when in tests, we use the `NoopTestWaiter` to noop the class' methods when run in production. This has a negligible impact on performance.
+
+### Using the `TestWaiter` class
+
+After building a test waiter, most users interact with a limited set of methods within this class, namely `beingAsync` and `endAsync`.
+
+The API used to signal whether an asynchronous operation has begun and ultimately ended is through the **_paired_** calls of `beginAsync` and `endAsync`: begin to denote the start of the asynchronous operation, and end to denote the end. Unique instances of async operations are identified using a `token` returned from `beginAsync`, which is subsequently provided to the `endAsync` call.
+
+To annotate the example provided above:
+
+```js
+import Component from '@ember/component';
+import { buildWaiter } from 'ember-test-waiters';
+
+// Creates a test waiter with the name 'friend-waiter' that
+// is usable by all instances of the `Friendz` component.
+let waiter = buildWaiter('friend-waiter');
+
+export default class Friendz extends Component {
+  didInsertElement() {
+    // Alerts the test waiter system that an async operation has started,
+    // storing the resulting unique token to be used to notify the test
+    // waiter system that the operation has ended.
+    let token = waiter.beginAsync();
+
+    someAsyncWork()
+      .then(() => {
+        //... some work
+      })
+      .finally(() => {
+        // Notifies the test waiter system that
+        // this unique async operation has ended.
+        waiter.endAsync(token);
+      });
+  }
+}
+```
+
+### `waitForPromise`
+
+The `waitForPromise` utility provides a convenience wrapper around the `TestWaiter` class for use with promises. It ensures the `endAsync` call is invoked in the `finally` of the configured promise.
```

Comment:
Can you add an example of how to use `waitForPromise`?

Reply:
Most definitely.

Reply:
Ya, good idea